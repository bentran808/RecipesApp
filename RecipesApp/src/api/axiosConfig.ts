import axios from 'axios';

export const axiosInstance = axios.create({
  baseURL: 'http://10.0.3.2:3000', //Change to http://10.0.2.2 for AVD and http://localhost for iOS
  headers: {
    'Content-Type': 'application/json'
  }
});
