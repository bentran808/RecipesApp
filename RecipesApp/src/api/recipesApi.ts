import { axiosInstance } from './axiosConfig';

export const recipesApi = {
  fetchCategoriesRequest() {
    return axiosInstance.get('categories?_embed=recipes');
  },
  fetchIngredientsRequest() {
    return axiosInstance.get('ingredients');
  },
  fetchRecipesRequest() {
    return axiosInstance.get('recipes?_expand=category');
  },
  fetchRecipesByCategoryIdRequest(id: number) {
    return axiosInstance.get(`recipes?_expand=category&categoryId=${id}`);
  },
  searchByCategoryNameRequest(keyword: string) {
    return axiosInstance.get(`categories?name_like=${keyword}&_embed=recipes`);
  },
  searchByRecipeNameRequest(keyword: string) {
    return axiosInstance.get(`recipes?title_like=${keyword}`);
  }
};
