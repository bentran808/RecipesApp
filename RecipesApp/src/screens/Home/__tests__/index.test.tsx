import { recipesApi } from 'api';
import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from 'screens/Home';
import { act, render } from '@testing-library/react-native';
import { recipe } from 'mocks';
import { Alert } from 'react-native';

describe('Home Screen', () => {
  let navigation: any;

  beforeEach(() => {
    navigation = {
      navigate: jest.fn(),
      setOptions: jest.fn(),
      openDrawer: jest.fn()
    };
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render correctly', () => {
    recipesApi.fetchRecipesRequest = jest.fn().mockReturnValue([recipe]);

    const tree = renderer.create(<HomeScreen navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should call function recipesApi.fetchRecipesRequest successful', () => {
    recipesApi.fetchRecipesRequest = jest.fn().mockReturnValue([recipe]);

    render(<HomeScreen navigation={navigation} />);

    expect(recipesApi.fetchRecipesRequest).toHaveBeenCalled();
  });

  test('should call function recipesApi.fetchRecipesRequest failed', () => {
    jest.spyOn(Alert, 'alert');
    recipesApi.fetchRecipesRequest = jest.fn().mockImplementation(() => {
      throw new Error('Network Error');
    });

    render(<HomeScreen navigation={navigation} />);

    expect(Alert.alert).toHaveBeenCalledWith('Fetch data failed');
  });

  test('should call function handleRefreshing successful', async () => {
    recipesApi.fetchRecipesRequest = jest.fn().mockReturnValue([recipe]);

    const { getByTestId } = render(<HomeScreen navigation={navigation} />);
    const list = getByTestId('recipesList');

    await act(async () => {
      list.props.onRefresh();
    });

    expect(recipesApi.fetchRecipesRequest).toHaveBeenCalled();
  });

  test('should call function handleRefreshing failed', async () => {
    jest.spyOn(Alert, 'alert');
    recipesApi.fetchRecipesRequest = jest.fn().mockImplementation(() => {
      throw new Error('Network Error');
    });

    const { getByTestId } = render(<HomeScreen navigation={navigation} />);
    const list = getByTestId('recipesList');

    await act(async () => {
      list.props.onRefresh();
    });

    expect(Alert.alert).toHaveBeenCalledWith('Fetch data failed');
  });
});
