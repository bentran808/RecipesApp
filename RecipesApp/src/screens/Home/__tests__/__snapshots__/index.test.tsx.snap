// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should render correctly 1`] = `
<View>
  <RCTScrollView
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    testID="recipesList"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
